src/controls/LevelingSensor/QskLevelingSensor.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include <QskControl.h>
+#include <memory>
+
+class QskScaleTickmarks;
+
+class QSK_EXPORT QskLevelingSensor : public QskControl
+{
+    Q_OBJECT
+    using Inherited = QskControl;
+
+  public:
+    QSK_SUBCONTROLS( OuterDisk, Horizon, TickmarksX, TickmarksXLabels, TickmarksY, TickmarksYLabels,
+        TickmarksZ, TickmarksZLabels )
+
+    using Tickmarks = QskScaleTickmarks;
+    using TickmarksLabels = QVector< QPair< qreal, QString > >;
+
+    explicit QskLevelingSensor( QQuickItem* parent = nullptr );
+    ~QskLevelingSensor();
+
+  public Q_SLOTS:
+    void setTickmarks( Qt::Axis axis, Tickmarks tickmarks );
+    void setTickmarksLabels( Qt::Axis axis, TickmarksLabels labels );
+    void setAngle( const QVector3D& degree );
+    void setAngle( Qt::Axis axis, float degree );
+
+    void setSubControlRotation( QskAspect::Subcontrol subControl, const QVector3D& degree );
+
+  Q_SIGNALS:
+    void anglesChanged( const QVector3D& degree );
+    void subControlRotationChanged( QskAspect::Subcontrol subControl, const QVector3D& degree );
+
+  public:
+    Q_REQUIRED_RESULT const Tickmarks& tickmarks( Qt::Axis axis ) const;
```

Comment:
Maybe just return the Tickmarks by value instead of const-ref, which is fine for Qt-style ref-counted classes?

Reply:
Well AFAIK the copy-of-write semantics can coexist with this approach. Personally I think this siganture 100% communicates what is hapening: The caller gets a readonly reference. Any further processing and manipulation can then trigger copy-on-write semantics.

In my opinion this whole copy-on-write semantics makes it even more complicated. Especially when mixing QSK and Qt containers. As a caller I always have to ask myself if an container/object is being copied and when. While the const ref approach clearly defines it.

Reply:
All that sounds good, but in all of Qt and QSkinny we return such classes by value, so IMO we should be consistent here too

Reply:
Well, for a user, who has been "socialized" from Qt APIs, there is a hidden message behind returning as reference: "the class being returned does not support copy-on-write".

Qt created their own type of containers back in the days, where there was no common STL and you could not rely on finding the same containers with each C++ compiler ( note that was also no std::string ). So it was actually mandatory for the cross platform concept. With Qt4 all containers have been redesigned from scratch - for some reasons ( I do not know ) not being replaced by the STL counter parts ( maybe because std::move did not yet exist ).

Unfortunately this lead to a situation, where a Qt addon project has to decide between the competing C++ and Qt concepts. For QSkinny I made the decision to go with Qt for the public API. The main reason for this decision is, that the majority of the QSkinny classes are derived from Qt classes and all parts of a class API should be consistent.

Considering designing Qsk classes, there is a simple rule of thumb: if a deep copy of a class is not cheap it is supposed to implement copy-on-write. Actually the idea of copy-on-write makes the life of a user simpler as he can copy around instances without thinking about extra costs.

Actually this worked perfectly fine until the range-based loops came up with C++11, where not so obvious "write" access of a container happens by accident.

Reply:
Maybe this and other information like the make_unique discussion is something we could gather on a list and publish it somewhere as coding guidelines or so?

Reply:
done